File: frontend/src/services/researchStrategy.ts
/** Confirmed research contracts, not trading policies or execution permissions. */
import { z } from "zod";
import { apiClient } from "./api";
import { policySchema } from "./portfolioRisk";
const n = z.number().finite();
export const parametersSchema = z.object({
  methods: z.array(z.enum(["peer_pe_annual@1", "fcff_proxy_dcf@1"])),
  peer_symbols: z.array(z.string()),
  peer_rationale: z.string(),
  discount_rate: n,
  growth_rate: n,
  terminal_growth: n,
  tax_rate: n,
  projection_years: z.number().int(),
  valuation_discount: n,
  earnings_decline_fraction: n,
  debt_to_fcf_limit: n,
  max_financial_age_days: z.number().int(),
  max_price_lag_sessions: z.number().int(),
  fcff_proxy_acknowledged: z.literal(true),
});
export type StrategyParameters = z.infer<typeof parametersSchema>;
const versionSchema = z.object({
  strategy_id: z.literal("fundamental-252@1"),
  version_id: z.string(),
  revision: z.number().int(),
  horizon: z.literal(252),
  horizon_unit: z.literal("XNYS_sessions"),
  benchmark: z.literal("SPY_total_return@1"),
  universe: z.literal("US_USD_nonfinancial_common_equity"),
  parameters: parametersSchema,
  cost_policy_revision: z.number().int(),
  cost_policy: policySchema,
  confirmed_at: z.string(),
  request_id: z.string(),
  input_hash: z.string(),
});
const stateSchema = z.object({
  revision: z.number().int(),
  active_version: z.string().nullable(),
  versions: z.array(versionSchema),
});
export type StrategyState = z.infer<typeof stateSchema>;
const inputSchema = z.object({
  snapshot_id: z.string(),
  evidence_id: z.string(),
  symbol: z.string(),
  metric: z.string(),
  value: n,
  unit: z.string(),
  period: z.string(),
  period_end: z.string(),
});
const valuation = z.object({
  method: z.enum(["peer_pe_annual@1", "fcff_proxy_dcf@1"]),
  status: z.enum(["available", "unavailable", "inapplicable"]),
  reasons: z.array(z.string()),
  inputs: z.array(inputSchema),
  assumptions: z.record(z.union([n, z.string()])),
  peer_multiple: n.nullable(),
  enterprise_value: n.nullable(),
  equity_value: n.nullable(),
  per_share_usd: n.nullable(),
  interpretation: z.literal(
    "conditional_model_estimate_not_verified_fair_value",
  ),
});
export const strategySummarySchema = z.object({
  reviews: z.array(
    z.object({
      review_id: z.string(),
      contract: versionSchema,
      symbol: z.string(),
      snapshot_id: z.string(),
      as_of: z.string(),
      purpose: z.literal("research"),
      status: z.enum([
        "research_only",
        "insufficient_evidence",
        "inapplicable",
      ]),
      errors: z.array(z.string()),
      valuations: z.array(valuation),
      monitoring: z.array(z.string()),
      checks: z.array(
        z.object({
          rule: z.string(),
          status: z.enum(["triggered", "not_triggered", "unavailable"]),
          value: n.nullable(),
          threshold: n,
          unit: z.literal("ratio"),
          evidence_ids: z.array(z.string()),
          reason: z.string(),
        }),
      ),
      conclusion: z
        .object({
          kind: z.literal("strategy_research"),
          symbol: z.string(),
          stance: z.enum(["bullish", "neutral", "bearish", "unknown"]),
          theses: z.array(
            z.object({
              text: z.string(),
              evidence_ids: z.array(z.string()),
              monitoring_rule: z.string(),
            }),
          ),
          scenarios: z.array(
            z.object({
              scenario: z.string(),
              condition: z.string(),
              probability: z.null(),
              probability_basis: z.literal("not_estimated"),
            }),
          ),
          report_markdown: z.string(),
          portfolio_action: z.null(),
          execution_intent: z.null(),
          prose_verification: z.literal("unverified"),
        })
        .nullable(),
      stale: z.boolean(),
      actionable: z.literal(false),
      prompt_versions: z.record(z.string()),
    }),
  ),
  errors: z.array(z.string()),
  legacy_strategy: z.literal(false),
  actionable: z.literal(false),
});
export type StrategySummary = z.infer<typeof strategySummarySchema>;
const headers = { "X-Financial-Agent-Local": "1" };
export async function getRunStrategy(
  runId: string,
  signal?: AbortSignal,
): Promise<StrategySummary | null> {
  return strategySummarySchema
    .nullable()
    .parse(
      (
        await apiClient.get<unknown>(
          `/api/portfolio/research-strategy/runs/${encodeURIComponent(runId)}`,
          { signal },
        )
      ).data,
    );
}
export async function getStrategy(
  signal?: AbortSignal,
): Promise<StrategyState> {
  return stateSchema.parse(
    (
      await apiClient.get<unknown>("/api/portfolio/research-strategy", {
        signal,
      })
    ).data,
  );
}
export async function confirmStrategy(
  parameters: StrategyParameters,
  expected_revision: number,
  cost_policy_revision: number,
  request_id: string,
): Promise<StrategyState> {
  return stateSchema.parse(
    (
      await apiClient.post<unknown>(
        "/api/portfolio/research-strategy",
        {
          parameters,
          expected_revision,
          cost_policy_revision,
          request_id,
          confirm: true,
        },
        { headers },
      )
    ).data,
  );
}
export async function deactivateStrategy(
  expected_revision: number,
): Promise<StrategyState> {
  return stateSchema.parse(
    (
      await apiClient.post<unknown>(
        "/api/portfolio/research-strategy/deactivate",
        { expected_revision },
        { headers },
      )
    ).data,
  );
}

File: backend/src/models/research_strategy.py
"""Confirmed research contracts and model-estimate receipts; never approval/execution."""

from datetime import date, datetime
from typing import Annotated, Literal

from pydantic import Field, model_validator

from .portfolio_risk import Number, Record, RiskPolicy, Symbol

Method = Literal["peer_pe_annual@1", "fcff_proxy_dcf@1"]


class StrategyParameters(Record):
    methods: list[Method] = Field(min_length=1, max_length=2)
    peer_symbols: list[Symbol] = Field(max_length=10)
    peer_rationale: str = Field(min_length=1, max_length=500)
    discount_rate: Annotated[Number, Field(gt=0, le=1)]
    growth_rate: Annotated[Number, Field(gt=-1, le=1)]
    terminal_growth: Annotated[Number, Field(gt=-1, lt=1)]
    tax_rate: Annotated[Number, Field(ge=0, le=1)]
    projection_years: int = Field(ge=1, le=20, strict=True)
    valuation_discount: Annotated[Number, Field(ge=0, lt=1)]
    earnings_decline_fraction: Annotated[Number, Field(gt=0, le=1)]
    debt_to_fcf_limit: Annotated[Number, Field(gt=0)]
    max_financial_age_days: int = Field(ge=1, le=730, strict=True)
    max_price_lag_sessions: int = Field(ge=0, le=5, strict=True)
    fcff_proxy_acknowledged: Literal[True]

    @model_validator(mode="after")
    def validate_definition(self) -> "StrategyParameters":
        if self.terminal_growth >= self.discount_rate:
            raise ValueError("Terminal growth must be below discount rate")
        if len(set(self.methods)) != len(self.methods):
            raise ValueError("Duplicate valuation methods are not triangulation")
        if len(set(self.peer_symbols)) != len(self.peer_symbols):
            raise ValueError("Duplicate peer instruments")
        return self


class StrategyVersion(Record):
    strategy_id: Literal["fundamental-252@1"] = "fundamental-252@1"
    version_id: str
    revision: int
    horizon: Literal[252] = 252
    horizon_unit: Literal["XNYS_sessions"] = "XNYS_sessions"
    benchmark: Literal["SPY_total_return@1"] = "SPY_total_return@1"
    universe: Literal["US_USD_nonfinancial_common_equity"] = (
        "US_USD_nonfinancial_common_equity"
    )
    parameters: StrategyParameters
    cost_policy_revision: int
    cost_policy: RiskPolicy
    confirmed_at: datetime
    request_id: str
    input_hash: str


class StrategyState(Record):
    revision: int = 0
    active_version: str | None = None
    versions: list[StrategyVersion] = Field(default_factory=list, max_length=100)


class ValuationInput(Record):
    snapshot_id: str
    evidence_id: str
    symbol: str
    metric: str
    value: Number
    unit: str
    period: str
    period_end: date


class ValuationResult(Record):
    method: Method
    status: Literal["available", "unavailable", "inapplicable"]
    reasons: list[str] = Field(default_factory=list)
    inputs: list[ValuationInput] = Field(default_factory=list)
    assumptions: dict[str, Number | str] = Field(default_factory=dict)
    peer_multiple: Number | None = None
    enterprise_value: Number | None = None
    equity_value: Number | None = None
    per_share_usd: Number | None = None
    interpretation: Literal["conditional_model_estimate_not_verified_fair_value"] = (
        "conditional_model_estimate_not_verified_fair_value"
    )


class Thesis(Record):
    text: str = Field(min_length=1, max_length=1200)
    evidence_ids: list[str] = Field(min_length=1, max_length=10)
    monitoring_rule: Literal["earnings_decline", "debt_to_fcf", "valuation_discount"]


class Forecast(Record):
    scenario: Literal["bull", "base", "bear"]
    condition: str = Field(min_length=1, max_length=1000)
    probability: None = None
    probability_basis: Literal["not_estimated"] = "not_estimated"


class StrategyConclusion(Record):
    kind: Literal["strategy_research"] = "strategy_research"
    symbol: Symbol
    stance: Literal["bullish", "neutral", "bearish", "unknown"]
    theses: list[Thesis] = Field(max_length=6)
    scenarios: list[Forecast] = Field(max_length=3)
    report_markdown: str = Field(min_length=1, max_length=20000)
    portfolio_action: None = None
    execution_intent: None = None
    prose_verification: Literal["unverified"] = "unverified"

    @model_validator(mode="after")
    def unique_scenarios(self) -> "StrategyConclusion":
        if len({s.scenario for s in self.scenarios}) != len(self.scenarios):
            raise ValueError("Duplicate conditional scenarios")
        return self


class StrategyConclusions(Record):
    conclusions: list[StrategyConclusion] = Field(max_length=100)


class MonitoringCheck(Record):
    rule: str
    status: Literal["triggered", "not_triggered", "unavailable"]
    value: Number | None
    threshold: Number
    unit: Literal["ratio"] = "ratio"
    evidence_ids: list[str] = Field(default_factory=list)
    reason: str


class StrategyReview(Record):
    review_id: str
    contract: StrategyVersion
    symbol: str
    snapshot_id: str
    as_of: datetime
    purpose: Literal["research"] = "research"
    status: Literal["research_only", "insufficient_evidence", "inapplicable"]
    errors: list[str]
    valuations: list[ValuationResult]
    monitoring: list[str]
    checks: list[MonitoringCheck] = Field(default_factory=list)
    conclusion: StrategyConclusion | None = None
    stale: bool = False
    actionable: Literal[False] = False
    prompt_versions: dict[str, str] = Field(default_factory=dict)


class StrategySummary(Record):
    reviews: list[StrategyReview]
    errors: list[str]
    legacy_strategy: Literal[False] = False
    actionable: Literal[False] = False
